File: library/src/main/java/io/github/pelmenstar1/rangecalendar/MathUtils.java
package io.github.pelmenstar1.rangecalendar;

import android.graphics.Color;

import androidx.annotation.ColorInt;

final class MathUtils {
    public static float lerp(float start, float end, float fraction) {
        return start + (end - start) * fraction;
    }

    public static int colorLerp(@ColorInt int start, @ColorInt int end, float fraction) {
        return Color.argb(
                (int)lerp(Color.alpha(start), Color.alpha(end), fraction),
                (int)lerp(Color.red(start), Color.red(end), fraction),
                (int)lerp(Color.green(start), Color.green(end), fraction),
                (int)lerp(Color.blue(start), Color.blue(end), fraction)
        );
    }
}

File: library/src/main/java/io/github/pelmenstar1/rangecalendar/MathUtils.kt
package io.github.pelmenstar1.rangecalendar

internal fun floorMod(x: Long, y: Long): Long {
    val r = x / y
    var aligned = r * y
    if (x xor y < 0 && aligned != x) {
        aligned -= y
    }
    return x - aligned
}

internal fun distanceSquare(x1: Float, y1: Float, x2: Float, y2: Float): Float {
    val xDist = x2 - x1
    val yDist = y2 - y1

    return xDist * xDist + yDist * yDist
}

internal fun lerp(start: Float, end: Float, fraction: Float): Float {
    return start + (end - start) * fraction
}

internal fun lerp(start: Int, end: Int, fraction: Float): Int {
    return start + ((end - start).toFloat() * fraction).toInt()
}